---
title: "Homework 3. data table"
author: "Amanda Jimenez"
date: "`r format(Sys.time(), '%d.%m.%Y')`"
output: html_notebook
---

# 81%  (56,00/69,00)


# Deadline for homework is Sunday,24.10.2021., 23.55 h

# Instructions and RULES:
+ Perform all of the analysis using only data.table. 
+ It is allowed to use functions that are present in base R (DO NOT forget them as you may need to search them and use them in this homework).
+ All the solutions which use for loops will not be graded. Use the functions from the apply family!!!
+ If certain task contains detailed instructions how to solve something, then those instructions must be followed or these tasks will not be graded.
+ Whenever you are asked to show a certain value or values as an answer and if it is not precisely stated in which format should the answer be (data.table or a vector) then both answers are correct.
+ Manual calculations and writing the answers directly on the code chunk will not be taken into consideration without the proper coding steps (and/or saved variable) that confirm this answer.


```{r}
library(data.table)
```

# Task 1. Analysis of dataset *flights* (44 pts)

Read in the flights data in flights variable from   "https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv". (Install curl package if you need to before doing this.)

```{r, message=FALSE}
library(data.table)
flights <- fread("https://raw.githubusercontent.com/Rdatatable/data.table/master/vignettes/flights14.csv")
flights
```

a1) Order the flights data table by carrier (ascenting order) then by decreasing distance, and save only hour as a vector in a variable var1a1. (1 point)
```{r, message=FALSE}
flights[order(carrier, -distance)] #by default order acts in ascending order
var1a1 <- flights[, hour]
var1a1
```
<span style="color:red"> 0/1 Using only order() does not permenantly change the ordering </span>


a2) Order the flights data table by carrier (ascenting order) then by decreasing distance, and save only hour as a data table with one column in a variable var1a2. (1 point)
```{r, message=FALSE}
flights[order(carrier, -distance)]
var1a2 <- data.table(flights[, list(hour)])
var1a2
```
<span style="color:red"> 0.75/1 Same mistake and defining again data table when subseting is not needed since it is a table. </span>


b1) Find the max flight air time for each carrier. copy the result to variable var1b1 (data.table with 2 columns, carrier and maxairtime, sorted by carrier. Do this in one command, no chaining, don't change the original data.table.). (1 point)  
```{r, message=FALSE}
var1b1 <- flights[order(carrier),.(max_air_time= max(air_time)), by=carrier]
var1b1
```
<span style="color:blue"> 1/1  </span>

b2) Find the max flight air time for each carrier. copy the result to variable var1b2 (data.table with 2 columns, carrier and maxairtime, sorted by maxairtime, from largest to smallest. Don't change the original data.table.). (1 point)  
```{r, message=FALSE}
var1b2 <- flights[ , .(max_air_time = max(air_time)), by=carrier][order(-max_air_time)]
var1b2
```
<span style="color:blue"> 1/1  </span>

c1) Count the number of flights that flew maximum air time for each carrier. Copy the data.table with carrier and numberofmaxflights to var1c1. Do this in one command, no addition of new columns and no chaining=2 points.  
```{r, message=FALSE}
var1c1 <- flights[, {max_air_time = max(air_time)
        .(Count_max=unlist(lapply(carrier, function(x) sum(air_time==max_air_time))))}, by=carrier]
```
<span style="color:red"> 1/2 better solution flights[, .(numberofmaxflights = sum(air_time == max(air_time)) ), by=carrier]. Same as Aitana. The more complicated the syntax, the more teamwork and copy-pasting is detected. Also the column name is supposed to be numberofmaxflights. </span>


c2) Then do this by addition of new column to flights and/or chaining in one command = 1 point save the result to var1c2 (copy!). If you add the column, remove it in the end of this exercise (end of c2 just remove the column/s if added from flights).  
```{r, message=FALSE}
var1c2 <- flights[order(air_time), Count_max := unlist(lapply(carrier, function(x) sum(air_time==max(air_time)))), by=carrier][, head(.SD[1,11],1), by=carrier]
var1c2
flights <- flights[,1:11]
```
<span style="color:blue"> 1/1 </span>


d1) Count how many flights had the same flight time for each carrier separately, ordered by carrier alphabetically and by decreasing flight time (carrier, air_time, N). copy the result to var1d1. Do this in one command, no chaining. (1 point)
```{r, message=FALSE}
var1d1 <- flights[order(carrier, -air_time), .N, by=.(carrier, air_time)]
var1d1
```

<span style="color:blue"> 1/1 </span>

d2) Get the data.table containing the same as previous (carrier, air_time, N), but only for the longest 2 flights for each carrier! copy the result to var1d2. Do this in one command, chaining if needed. (1 point)  
```{r, message=FALSE}
var1d2 <- flights[order(carrier, -air_time), .N, by=.(carrier, air_time)][, head(.SD,2), by=carrier]
#var1d2 <- flights[order(carrier, -air_time), head(.SD,2), by=carrier][, .N, by=.(carrier, air_time)]
var1d2
```
<span style="color:blue"> 1/1 </span>

d3) Get the data.table containing the same as previous (carrier, air_time, N), but only for the longest and shortest flights for each carrier! copy the result to var1d3. Do this in one command, chaining if needed. (1 point)  
```{r, message=FALSE}
var1d3 <- flights[order(carrier, -air_time), .N, by=.(carrier, air_time)][, .SD[c(1, .N)], by=carrier]
var1d3
```
<span style="color:blue"> 1/1 </span>

e) Get the name of the carrier who has the largest difference between longest and shortest flight air time! copy the result to var1e (only the carrier name as an element of vector). Do this in one command, chaining if needed, do not change the original data.table! (3 points)  
```{r, message=FALSE}
var1e <- as.character(flights[, .(air_time_diff=max(air_time)-min(air_time)), by=carrier][order(air_time_diff), head(.SD[,1],1)])
var1e
```
<span style="color:red"> 2/3 Wrong ordering. </span>

f) Get the name of the carrier who has the largest mean air time. One command, chaining allowed, don't change the original data.table. Save the name alone in variable var1f (still the same command.). (1 point)  
```{r, message=FALSE}
var1f <- as.character(flights[, .(mean_air_time=mean(air_time)), by=carrier][order(mean_air_time), head(.SD[,1],1)])
var1f
```
<span style="color:blue"> 1/1 Again wrong ordering.  </span>

g1) Write a function that will calculate the squared sum of differences between each element and the mean value of all elements, call the function myRss. The function should take one vector as an argument and return numeric value. Use this function to calculate the value for air time for each carrier, name the column rss, dont change the original table! one command, no chaining. (copy to var1g1, 2 points).  
```{r, message=FALSE}
myRSS <- function(vector){
  rss <- sum((vector-mean(vector))^2)
  rss
}
var1g1<- flights[,.(rss=( myRSS(air_time))),
by=carrier]
var1g1
```

<span style="color:blue"> 2/2 </span>


g2) Do as the previous but for air_time and arr_delay columns in the same command, no chaining, no change to original data.table (var1g2, 1 point)!   
```{r, message=FALSE}
var1g2<-copy(flights[,lapply(.SD,myRSS),by="carrier",.SDcols=c("air_time","arr_delay")])
var1g2
```
<span style="color:blue"> 1/1 </span>

g3) Do as the previous but for all numeric columns in the same command, no chaining, no change to original data.table (var1g3, 2 points)!   
```{r, message=FALSE}
var1g3<-flights[,lapply(.SD,myRSS),by="carrier",.SDcols=sapply(flights,is.numeric)]
var1g3 
```
<span style="color:blue"> 2/2 </span>

h1) Assign a new column to flights: speed which will have speed calculated for each flight. Do it in a data.table fashion. copy the resulting flights data.table to myFlights1 variable. (1 point)
```{r, message=FALSE}
myFlights1 <- copy(flights)
myFlights1 <- copy(myFlights1[, speed := distance/(air_time/60)][])
myFlights1
# Because of the next task, I assume this speed had to be an integer, so:
myFlights1 <- copy(myFlights1[, speed := as.integer(distance/(air_time/60))][])
myFlights1
```
<span style="color:blue"> 1/1  </span>

h2) Convert speed and distance variables to type double. do it in data.table fashion, in one command. copy the resulting data.table to myFlights2 variable. Do this withouth lapply. (1 point). 
```{r, message=FALSE}
myFlights2 <- copy(myFlights1[, ":=" (distance=as.double(distance), speed=as.double(speed))][])
myFlights2
```
<span style="color:blue"> 1/1 </span>

h3) Again do the RSS calculation as in var1g1 but for all double columns in the same command, no chaining, no change to original data.table (var1g4, 1 point)!   
```{r, message=FALSE}
var1g4<-myFlights2[,lapply(.SD,myRSS),by="carrier",.SDcols=sapply(myFlights2,is.double)]
var1g4
```
<span style="color:blue"> 1/1 </span>

i) Save the names of the columns which are of type double to thosecolumns variable.(1 point)
```{r, message=FALSE}
thosecolumns <- names(myFlights2)[sapply(myFlights2, is.double)]
thosecolumns
```
<span style="color:blue"> 1/1 </span>

j) Convert thosecolumns back to type integer. do it in data.table fashion, in one command.  copy the resulting data.table to myFlights3 variable. But now do this with lapply.(2 points).
```{r, message=FALSE}
myFlights3<-copy(myFlights2[,(thosecolumns):=(lapply(.SD,as.integer)), .SDcols=thosecolumns])
myFlights3
```
<span style="color:blue"> 2/2 </span>

k) For air_time column, calculate the difference between the second and first observation for each carrier, in flights data frame ordered by carrier and then by decreasing air_time. In the same command, no chaining, no change to original data.table, no additional function defined outside (var1k, 2 points)!
```{r, message=FALSE}
var1k<- flights[, .(Difference=unlist(lapply(.SD[c(2:1)], function(vector){
  setorder(flights,carrier,-air_time); diff(vector)}))),
  by="carrier", .SDcols="air_time"]
var1k
#var1k_old <- copy(var1k)
var1k_old
```
<span style="color:red"> 1/2 After the second running when setorder gets noted, you get the almost correct calculations but first - second. One of solutions is flights[order(carrier, -air_time) ,air_time[2]-air_time[1]  , carrier] or flights[order(carrier, -air_time), .SD[1:2, diff(air_time)], by=carrier]. </span>

l) For all integer columns, calculate the difference between the second and first observation for each carrier. In the same command, no chaining, no change to original data.table, no additional function defined outside (var1l, 2 points)!
```{r, message=FALSE}
var1l<- flights[,lapply(.SD[c(2:1)], function(vector){
  diff(vector)}), by="carrier", .SDcols=sapply(flights, is.integer)]
var1l
```
<span style="color:blue"> 2/2 Wrong order. </span>

m) Copy the flights to myFlights. Add a new column, rss_air_time which will have myRss of air_time saved in it, to myFlights. do not change the flights variable, dont use chaining, do it in one command, (myFlights assignment is separate command.) copy the result to variable myFlights1
```{r, message=FALSE}
myFlights <- copy(flights)
myFlights1 <- myFlights[, ":=" (rss_air_time=myRSS(myFlights$air_time))]
myFlights1
```
<span style="color:red"> 0.75/1 myFlights[, rss_air_time := myRss(air_time)]. The correct data.table way. </span>

n) Make a vector with all the names of the columns with integer values, save it in intnames variable. Use the **paste** function to make a vector which will have all the names of integer columns converted to rss_columnname (so for air_time it will be rss_air_time, dor hour it will be rss_hour, ...) (2 points) (No need to do this with data table). save the vector as rssnames variable.
```{r, message=FALSE}
col_names <- flights[,names(which(sapply(flights,is.integer)))]
rssnames <- paste("rss",col_names,sep="_")
rssnames
```
<span style="color:blue"> 2/2 </span>

o) Use the previously created vectors to add multiple new columns to data.table. The names of the columns you will add will be rssnames. And the values will of course be myRss(column), where column is every column from intnames. Do it in one command, data.table style, no chaining. Copy the result to myFlights3 variable. (2 points).
```{r, message=FALSE}
myFlights <- copy(flights)
myFlights3 <- myFlights[,c(rssnames):=(lapply(.SD,myRSS)),.SDcols=col_names]
myFlights3
```
<span style="color:blue"> 2/2 </span>

p) Go back to flights data.table. Calculate and return (do not change the data.table) the mean, and sd for air_time for flights which travel between JFK and MIA. Do this by each carrier, in one line, no chaining. Do not rename the columns to anything, copy the result to myFlights4.(1 points)  
```{r, message=FALSE}
myFlights4 <- flights[((origin=="JFK" & dest=="MIA") | (origin=="MIA" & dest=="JFK")),  .(mean=mean(air_time), sd=sd(air_time)), by=carrier]
myFlights4
```

<span style="color:blue"> 1/1 </span>

r)  Which journey (from -> to, regardless of the carrier) was delayed in arrival the most times? (var1r, 1 pts)
```{r, message=FALSE}
var1r <- flights[, .(number_of_delays=sum(sign(arr_delay)>0)), by=.(origin, dest)][number_of_delays==max(number_of_delays), .(from=origin, to=dest)]
var1r
```
<span style="color:blue"> 1/1 </span>


s)  Which journey (from -> to, regardless of the carrier) was delayed in arrival the most in total minutes? (var1s, 1 pts)
```{r, message=FALSE}
# This is the one with de longest delay in minutes:
var1s <- flights[arr_delay==max(arr_delay),  .(from=origin, to=dest),]
var1s
# This is the one that, adding up all the delays, has the most delay minutes in total:
var1s <- flights[, .(total_mins_delay=sum(arr_delay)), by=.(origin, dest)][total_mins_delay==max(total_mins_delay), .(from=origin, to=dest)]
var1s
```
<span style="color:blue"> 1/1. Second one correct </span>

t) Are flights longer depending on time of the day? (Divide the day into : night (from 22->6 hours), morning (6->12, afternoon 12->16, evening
16->22 )) (var1t, 3)
```{r, message=FALSE}
myFlights <- copy(flights)
flights_daytime <- myFlights[, daytime := ifelse(hour>=22 & hour<6, "night",
                             ifelse(hour>=6 & hour<12,  "morning",
                              ifelse(hour>=12 & hour<16, "afternoon", "evening")))][]
var1t <- flights_daytime[, .(time_of_flight = mean(air_time)), by="daytime"]
var1t
# from this I can conclude that the flights are longer on the mornings. There are no flights during the night.
```
<span style="color:red"> 2/3. The dividing is not correct. Hour 5 should be in night and not evening category. This condition is not correctly defining the night category hour>=22 & hour<6. BEcause hour of 1 shall never be larger than 22.  </span>

u) For which time of the day are flights delayed in departure the most and for which the least? What about arrival delays? (var1u, 2)
```{r, message=FALSE}
myFlights <- copy(flights)
flights_daytime <- myFlights[, daytime := ifelse(hour>=22 & hour<6, "night",
                             ifelse(hour>=6 & hour<12,  "morning",
                              ifelse(hour>=12 & hour<16, "afternoon", "evening")))][]
var1u <- flights_daytime[, .(mean_dep_delay=mean(dep_delay), mean_arr_delay=mean(arr_delay)), by="daytime"][order(mean_dep_delay, mean_arr_delay)] # number_of_delays=sum(sign(column)>0) 

#Most delay in the departure time in the evening, least delay in the departure time in the morning. Same for the arrival delay. These results are taking into account the mean minutes of delay, not the number of flights delayed in each period (which could be calculated changing the function, as it says in the comment next to var1u variable)
```
<span style="color:red">  1/2  The question is "the most and for which the least" in delay which need to be answered with a code. </span>

v) Answer u) but take into account the day of the week as well. Use package function **wday** from package **lubridate** to answer this question. (For which day of the week and which time of the day are flights delayed the least, and for which the most?) (var1u, 2)
```{r, message=FALSE}
#install.packages("lubridate")
#library("lubricate")
flights_daytime2 <- copy(flights_daytime)

var1u <- flights_daytime2[, date := paste(year,month,day, sep="-")][, day_of_the_week := wday(date)][, .(mean_dep_delay = mean(dep_delay), mean_arr_delay = mean(arr_delay)), by=.(day_of_the_week, daytime)][order(mean_dep_delay)]
var1u
# wday takes Sunday as the day 1 by default, so the most departure delay minutes accumulate on Thursday's afternoons, and the least on Monday's mornings.
# The most arrival delay minutes accumulate on exactly same days and same periods of the day, and the same happen with the least delay minutes accumulated.

# Once again, my calculations are based on the mean minutes of delay (advanced if the number is negative) throughout the daytimes and the days of the week. I also thought it could be done with the number of flights delayed (not taking into account the minutes) in the same periods of time, using number_of_delays=sum(sign(column)>0)  instead of the mean.
```
<span style="color:red">  2/2  Same comment as above.  </span>


# Task 2. Taxonomic analysis of sponge genomes (25 pts)

## Short introduction
Program DIAMOND is a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data. When combined with program MEGAN, it becomes an indispensable tool for taxonomic and functional analysis of metagenomes. More information can be found at https://link.springer.com/protocol/10.1007/978-1-4939-9074-0_19. Not only is it useful for analyzing metagenomes but this workflow can also be used in detection of contaminants in newly assembled genomes of Eukaryotes. In this task you will have to analyze DIAMOND results for the genomes of several sponges.

Before you start solving this task, you need to understand how BLAST and DIAMOND work since you are analyzing its output.
Answer these questions before solving the tasks. (1)

What is DIAMOND and why is it similar to BLAST?

Both are Average Amino Acid Identity (AAI) tools to compare the similarity between two genomes. So both are using for protein alignment (BLASTx in this case), but DIAMOND is much faster with a really similar sensibility, so it has become more popular to analize larger genomes.

What is the output format and what do the columns represent (focus on columns present in table *spongeDiamond* in step 0. in the following task)?
The output format is a datatable.
- First column is the commong name for each type of sponge
- qseqid is Query Seq-id
- qlen is Query sequence length
- sseqid means Subject Seq-id
- slen means Subject sequence length
- staxids means unique Subject Taxonomy ID(s)
- separated by a ';' (in numerical order)
- stitle means Subject Title
- qcovhsp means Query Coverage Per HSP
- length means Alignment length
- pident means Percentage of identical matches
- mismatch means Number of mismatches
- qstart means Start of alignment in query
- qend means End of alignment in query
- sstart means Start of alignment in subject
- send means End of alignment in subject
- evalue means Expect value
- bitscore means Bit score
- Superkingdom is the domain of the species
- phylum, class, order, family, genus and species is the clasification of each sponge, from the broadest step od the classification (phylum) to the more specific (species)
Every row is a contig, with its unique ID in the qseqid column, but one contig can appear multiple times since it can have a match multiple times.

<span style="color:blue"> 1/1 pts </span>

## Tasks
0.step: Import the file **SpongeDiamond_algorithms2021.RDS** and examine the table.
```{r}
spongeDiamond <- readRDS("SpongeDiamond_algorithms2021.RDS")
spongeDiamond
```

a) How many sponges do you have in your table? (1)
```{r}
var2a <- spongeDiamond[, uniqueN(sponge)]
var2a
```
<span style="color:green"> 1/1 pts </span>

b) Which sponge (or sponges if more have the same result) has the most contigs? (1)
```{r}
var2b <- spongeDiamond[, .N, by=.(sponge, qseqid)][, .N, by="sponge"]
var2b
#Amphimedon is the one with more different contigs.
```
<span style="color:red"> 0.75/1 pts Your results says otherwise. Oscarella had the most contigs (from your result) </span>

c) Which sponge (or sponges if more have the same result) has the shortest median length of contigs? (1)
```{r}
var2c <- spongeDiamond[, .(median_qlength = median(qlen)), by=.(sponge)][order(median_qlength)]
var2c
#Oscarella
```
<span style="color:red"> 0.5/1 pts You have many duplicates of your contings and you have to perform a filtering to obtain unique values prior to calculating the median.  </span>

d) Which contig has the highest and which contig the lowest number of mismatches in total for each sponge? (1) 
```{r}
var2d <- spongeDiamond[order(-mismatch), .SD[c(1, .N), c(1,10)], by=.(sponge)]
var2d
```
<span style="color:red"> 0.75/1 pts In total refers to the sum but very nice clean subsetting :D. </span>

e) Which pair of superkingdom-family occurred most time for Amphimedon? (2)
```{r}
var2e <- spongeDiamond[sponge=="Amphimedon", .N, by=.(superkingdom, family)][order(-N)]
var2e
# the pair Eukaryota	Niphatidae
```
<span style="color:green"> 2/2 pts </span>

f) Select protein-sponge pair with the highest aligned lengths for each pair and calculate the difference between the pairs with the highest sum of length and lowest sum of lengths. (2)
```{r}
# I don't really know what to "sum" here (it could refer to the function that makes the summatory, or simply as a way to refer the value, the quantity), so this is my interpretation taking the max length of each pair, and obtaining the difference between the pair with the highest highest length  and the pair with the lowest highest length:
var2f1 <- spongeDiamond[, .(highest_aligned=max(length)), by=.(stitle, sponge)][order(-highest_aligned), .(difference=head(highest_aligned, 1)-tail(highest_aligned,1))]
var2f1

# Another interpretation could be actually using the sum function, so I make the summatory of the allignment lengths for each pair, then take the highest of those and I rest the lowest value for that summatory within each pair:
var2f2 <- spongeDiamond[, .(highest_aligned=sum(length)), by=.(stitle, sponge)][order(-highest_aligned), .(difference=head(highest_aligned, 1)-tail(highest_aligned,1))]
var2f2
```
<span style="color:green"> 2/2 pts </span>

g) Annotate every query contigs for each sponge in the original cleaned *spongeDiamond* table in the following way: If best 3 results (by eval) for each contig belongs to the same superkingdom, annotate that contig as that superkingdom. (2) Note: Select smaller sample size on which you will develop your code and then run it on the whole dataset if you are having troble with the running time or crashing. 
```{r}
var2g1 <- annotated <- unique(spongeDiamond[order(evalue), .SD[1:3,c(15,17)], by="qseqid"][, .SD[length(unique(superkingdom))==1, 2], by="qseqid"])
var2g1
```
 <span style="color:red"> 1/2 pts It is needed to be done on group by= .(sponge, qseqid) because some contigs in different sponges have the same id. Like scaffold_0001 in Tethya and Ephydatia.. In this annotation you have lost some contigs such as "Contig13327" because you are doing subsetting and annotating based on the set criteria. You should have kept those contigs and label them as not annotated  (or NA) because they are also an important part of your data that you want to report. </span>  
  
  + What is the difference in the number of contaminants (sequences belonging to superkingdom Bacteria) between the sponge with the highest number of contaminants and the rest ? Show the results as data table with sponges and differences (3)

```{r}
var2g2 <- spongeDiamond[superkingdom=="Bacteria", .N, by="sponge"][order(-N), .(sponge, diff_contaminant=unlist(lapply(.SD[,2],function(x){ x[1]-x})))]
var2g2
```
<span style="color:red"> 2/3 pts Contaminants are contings belonging to Bacteria superkingdom. Here you have forgotten the unique since there are multiple same contings that are from bacteria.  </span>
 
h) Add a column "pidentGrp" that will represent identity groups for each row : cut the “pident” value to bins of 0-10%, 10-20%, 20-30%,…90-100%. 
Add a column "maxLength", that will represent maximum of aligned length for each contig in sponge- this is the largest match possible for each contigs and all hits. Add another column "percMaxLen" that will say what percentage of maxLength the row has. (3)
```{r}
var2h <- spongeDiamond[, pidentGrp := ifelse(pident>=0 & pident<10, "0-10%",
                             ifelse(pident>=10 & pident<20, "10-20%",
                              ifelse(pident>=20 & pident<30, "20-30%", 
                                ifelse(pident>=30 & pident<40, "30-40%",
                                  ifelse(pident>=40 & pident<50, "40-50%",
                                    ifelse(pident>=50 & pident<60, "50-60%",
                                      ifelse(pident>=60 & pident<70, "60-70%",
                                        ifelse(pident>=70 & pident<80, "70-80%",
                                          ifelse(pident>=80 & pident<90, "80-90%",
                                            "90-100%")))))))))][]
spongeDiamond[, ':='(maxLength=max(length), percMaxLen=(length/max(length))*100), by="qseqid"][]
var2h
```
<span style="color:red"> 2.5/3 pts  What would you do if the group percentages had to be of 1%? See function cut or what does %/% do and how can you use them. </span>

i) In this part functions rleid and shift might be useful to you. What do they do? (2 points). 
```{r}
?rleid
# A convenience function for generating a run-length type id column to be used in grouping operations. It accepts atomic vectors, lists, data.frames or data.tables as input. At times aggregation (or grouping) operations need to be performed where consecutive runs of identical values should belong to the same group. This function allows to generate "run-length" groups directly.
?shift
#lead or lag vectors, lists, data.frames or data.tables implemented in C for speed. shift accepts vectors, lists, data.frames or data.tables. It always returns a list except when the input is a vector and length(n) == 1 in which case a vector is returned, for convenience.

# In the case of datatable.shift(): Produce a column obtained from col shifting it n rows forward.
# The shift amount, n, can be both positive and negative. If positive, a “lag” column is created, if negative it will be a “lead” column.
#The shifted column will have the same number of rows as the original column, with n observations in the beginning becoming missing, and n observations at the end discarded.
```

<span style="color:red"> 1/2 pts Write in your own words and not just copy paste from the manual. If you have given an example and explained it that would be fine.  </span>  

  + Read in file with one whole contig for Tethya (**contigTethya.txt**) find which "class" appears mosttimes. (save the Tethya genome to new a variable "genomeTethya" and continue to work with it). For example if "class" repeat is Demospongiae, Demospongiae, Leptocardii, Leptocardii, Leptocardii, Leptocardii then the class Leptocardii appears most times consecutively, 3 time to be precise. If multiple class appear the same times, report all of them. (4 pts)
```{r}
genomeTethya <- data.table::fread("contigTethya.txt")
genomeTethya
var2i <- genomeTethya[,.(class, assigned_type=rleid(class))][, .(class=unique(class), sum_types=.N), by="assigned_type" ][order(-sum_types), .SD[1, 2:3]]
var2i
```
<span style="color:green"> 4/4 pts </span>

  +  In which positions are those Families (do the analysis on the variable "longestTethya")? Save the result as vector. (2 points)
```{r}
longestTethya <- as.vector(unlist(genomeTethya[, .(class, row_id=.I)][class=="Mammalia", .SD[,2]]))
longestTethya
# I undertood that the shift function serves to move entire rows up and down, or to operate with previous or next row, but I didn't really needed it for these tasks.
```

<span style="color:red"> 0/2 pts  This continues to the previous task and ask for "those" families which appear consecutively most times. </span>
